### PR TITLE
OpenStack: add clusterOSImage validations

### PR DESCRIPTION
**Pull request by Fedosin**
This commit adds validations that custom Glance image specified by `clusterOSImage` config parameter exists and has Active status.

### CODE REVIEW

File: pkg/asset/installconfig/openstack/validation/cloudinfo.go [hunk 1]
```diff
@@ -151,3 +170,24 @@ func (ci *CloudInfo) getNetwork(networkName string) (*networks.Network, error) {
 
 	return network, nil
 }
+
+func (ci *CloudInfo) getImage(imageName string) (*images.Image, error) {
+	if imageName == "" {
+		return &images.Image{}, nil
```
Comment (pierreprinetti) [inline pkg/asset/installconfig/openstack/validation/cloudinfo.go:176]: Question: Why do you return a pointer to an empty image? I expected to see `return nil nil` here

Comment (Fedosin) [reply pkg/asset/installconfig/openstack/validation/cloudinfo.go:176]: This is an absolutely useless check and should be removed. when we call `getImage` we check that the image name is non-empty.
Thank you for pointing at this!

File: pkg/asset/installconfig/openstack/validation/cloudinfo.go [hunk 2]
```diff
@@ -88,6 +98,14 @@ func (ci *CloudInfo) collectInfo(ic *types.InstallConfig) error {
 		return errors.Wrap(err, "failed to fetch platform flavor info")
 	}
 
+	// Fetch the image info if the user provided a Glance image name
+	imagePtr := ic.OpenStack.ClusterOSImage
+	if imagePtr != "" {
+		if _, err := url.ParseRequestURI(imagePtr); err != nil {
+			ci.OSImage, err = ci.getImage(imagePtr)
```
Comment (abhinavdahiya) [inline pkg/asset/installconfig/openstack/validation/cloudinfo.go:105]: this err here is not handled

File: pkg/asset/installconfig/openstack/validation/platform.go [hunk 3]
```diff
@@ -127,3 +132,28 @@ func validateFloatingIPs(p *openstack.Platform, ci *CloudInfo, fldPath *field.Pa
...
 }
+
+// validateExternalNetwork validates the user's input for the clusterOSImage and returns a list of all validation errors
+func validateClusterOSImage(p *openstack.Platform, ci *CloudInfo, fldPath *field.Path) (allErrs field.ErrorList) {
+	if p.ClusterOSImage == "" {
+		return
+	}
+
+	// If the user input is an http address, then skip the validation
+	if _, err := url.ParseRequestURI(p.ClusterOSImage); err == nil {
```
Comment (abhinavdahiya) [inline pkg/asset/installconfig/openstack/validation/platform.go:143]: this comment does not match what is being done here, it's checking for a valid URI which includes any scheme not just http.

if you support http and https only, add validation for that here.

File: pkg/asset/installconfig/openstack/validation/platform.go [hunk 4]
```diff
@@ -127,3 +132,28 @@ func validateFloatingIPs(p *openstack.Platform, ci *CloudInfo, fldPath *field.Pa
...
+		return
+	}
+
+	// Image should exist
```
Comment (abhinavdahiya) [inline pkg/asset/installconfig/openstack/validation/platform.go:147]: can you expand here why image should exist or where it should exist?

File: pkg/asset/installconfig/openstack/validation/platform.go [hunk 5]
```diff
@@ -127,3 +132,28 @@ func validateFloatingIPs(p *openstack.Platform, ci *CloudInfo, fldPath *field.Pa
...
+	if ci.OSImage == nil {
+		allErrs = append(allErrs, field.NotFound(fldPath.Child("clusterOSImage"), p.ClusterOSImage))
+		return allErrs
+	}
+
+	// Image should be active
+	if ci.OSImage.Status != images.ImageStatusActive {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("clusterOSImage"), p.ClusterOSImage, "image is not active"))
```
Comment (abhinavdahiya) [inline pkg/asset/installconfig/openstack/validation/platform.go:155]:
```suggestion
		allErrs = append(allErrs, field.Invalid(fldPath.Child("clusterOSImage"), p.ClusterOSImage, "image must be active but it was %s"))
```

Comment (abhinavdahiya) [reply pkg/asset/installconfig/openstack/validation/platform.go:155]: provide details on the failed image.

File: pkg/asset/installconfig/openstack/validation/platform_test.go [hunk 6]
```diff
@@ -249,3 +250,96 @@ func TestOpenStackPlatformValidation(t *testing.T) {
...
 }
+
+func TestClusterOSImage(t *testing.T) {
+	cases := []struct {
+		name           string
+		platform       *openstack.Platform
+		cloudInfo      *CloudInfo
+		networking     *types.Networking
+		expectedError  bool
+		expectedErrMsg string // NOTE: this is a REGEXP
```
Comment (abhinavdahiya) [inline pkg/asset/installconfig/openstack/validation/platform_test.go:261]: `expectedError  bool` is just duplication of the requirement from `expectedErrMsg string`, if expectedErrMsg is not empty the error must match that regex else when empty there should be no error.

Comment (Fedosin) [reply pkg/asset/installconfig/openstack/validation/platform_test.go]: removed it

File: pkg/asset/installconfig/openstack/validation/platform.go [hunk 7]
```diff
@@ -127,3 +132,33 @@ func validateFloatingIPs(p *openstack.Platform, ci *CloudInfo, fldPath *field.Pa
...
+	if p.ClusterOSImage == "" {
+		return
+	}
+
+	// For URLs we support only 'http(s)' and 'file' schemes
+	if uri, err := url.ParseRequestURI(p.ClusterOSImage); err == nil {
+		switch uri.Scheme {
+		case "http", "https", "file":
+		default:
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("clusterOSImage"), p.ClusterOSImage, fmt.Sprintf("URL scheme should be either http(s) or file but it is '%v'", uri.Scheme)))
```
Comment (abhinavdahiya) [inline pkg/asset/installconfig/openstack/validation/platform.go:147]: Go has fmt directive %q which will quote the string

File: pkg/asset/installconfig/openstack/validation/platform.go [hunk 8]
```diff
@@ -127,3 +132,33 @@ func validateFloatingIPs(p *openstack.Platform, ci *CloudInfo, fldPath *field.Pa
...
+
+	// Image should exist in OpenStack Glance
+	if ci.OSImage == nil {
+		allErrs = append(allErrs, field.NotFound(fldPath.Child("clusterOSImage"), p.ClusterOSImage))
+		return allErrs
+	}
+
+	// Image should have "active" status
+	if ci.OSImage.Status != images.ImageStatusActive {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("clusterOSImage"), p.ClusterOSImage, fmt.Sprintf("OS image must be active but its status is '%s'", ci.OSImage.Status)))
```
Comment (abhinavdahiya) [inline pkg/asset/installconfig/openstack/validation/platform.go:160]: Go has fmt directive %q which will quote the string

**Pull request merged**